### PR TITLE
[SNOW-870373] Enable JMX metrics for Snowpipe Streaming

### PR DESCRIPTION
**Pull request by sfc-gh-rcheng**
Tracks the offsetPersistedInSnowflake, processedOffset and latestConsumerOffset for Snowpipe Streaming. This PR only enables the JMX metrics, further metrics may be added in the future as needed, please comment any metrics you think would be useful to track!

image

### CODE REVIEW

**Comment by sfc-gh-xhuang** [comment]
Putting reminder here to update our docs when this is available in the next release:
https://docs.snowflake.com/en/user-guide/kafka-connector-monitor

File: src/main/java/com/snowflake/kafka/connector/internal/metrics/MetricsUtil.java [hunk 1]
```diff
@@ -70,6 +70,14 @@ public class MetricsUtil {
...
   public static final String PURGED_OFFSET = "purged-offset";
 
+  /**
+   * See {@link com.snowflake.kafka.connector.internal.streaming.TopicPartitionChannel} for offset
+   * description
+   */
+  public static final String OFFSET_PERSISTED_IN_SNOWFLAKE = "persisted-in-snowflake-offset";
+
+  public static final String LATEST_CONSUMER_OFFSET = "latest-consumer-offset";
+
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/metrics/MetricsUtil.java:80]: They're really useful for our internal debugging, do you think they can be send to Snowhouse as well?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 2]
```diff
@@ -1303,4 +1317,45 @@ public String toString() {
...
   }
+
+  /**
+   * Registers all the Metrics inside the metricRegistry.
+   *
+   * @param channelName channelName
+   * @param metricsJmxReporter wrapper class for registering all metrics related to above connector
+   *     and channel
+   */
+  public void registerChannelJMXMetrics(
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1328]: Could we add latency related metrics similar to https://docs.snowflake.com/en/user-guide/kafka-connector-monitor#category-latencies?

**Review by sfc-gh-tzhang — APPROVED** [review]
lgtm!

File: src/main/java/com/snowflake/kafka/connector/internal/metrics/MetricsUtil.java (hunk 1 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/metrics/MetricsUtil.java:80]: 100% we can send it through the telemetry service, we currently don't have any telemetry coming out streaming specifically (similar to pipe_usage). planning to add it in a follow up pr if i have time. [jira](https://snowflakecomputing.atlassian.net/browse/SNOW-899866)

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java (hunk 2 above)
Comment (sfc-gh-rcheng) [reply src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:1328]: yes, i'm planning to add that in a follow up pr. wanted to keep this one small to just enable JMX, and then add latencies and buffer thresholds later

**Pull request merged**